eclass/dkms.eclass

```diff
@@ -0,0 +1,527 @@
+# }
+# @CODE
+#
+# Note that due to the inherit order the src_install and pkg_postinst
+# phase functions may have to be defined explicitly.
+#
+# @EXAMPLE:
+#
+# A more complex example is the case of an ebuild that is currently
+# inheriting linux-mod-r1, but is not using any of its phase
+# functions. In this case there is usually no modlist for
+# dkms_autoconf to convert into a DKMS configuration file.
+# Instead the ebuild must utilize a dkms.conf provided by upstream
+# in the sources, or alternatively create one from scratch and
+# include it in FILESDIR.
+#
+# Tip: Check if there is a rpm/deb spec or similar script that can
+# create a dkms.conf to find a hint of what it should look like and
+# where it should be created for this particular package.
+#
+# @CODE
+# inherit dkms linux-mod-r1
+#
+# src_prepare() {
+#   default
+#   sed -e "s/@VERSION@/${PV}/" -i modules/dkms.conf || die
+# }
+#
+# src_compile() {
+#     if use dkms; then
+#         dkms_gentoofy_conf modules/dkms.conf
+#     else
+#         emake "${MODULES_MAKEARGS[@]}" modules
+#     fi
+# }
+#
+# src_install() {
+#     if use dkms; then
+#         dkms_dopackage modules
+#     else
+#         linux_domodule modules/mymodule.ko
+#         modules_post_process
+#     fi
+#     einstalldocs
+# }
+#
+# pkg_postinst() {
+#     dkms_pkg_postinst
+# }
+# @CODE
+
+case ${EAPI} in
+	8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI:-0} not supported" ;;
+esac
+
+if [[ -z ${_DKMS_ECLASS} ]]; then
+_DKMS_ECLASS=1
+
+inherit linux-mod-r1
+
+IUSE="dkms"
+
+RDEPEND="dkms? ( sys-kernel/dkms ${BDEPEND} )"
+IDEPEND="dkms? ( sys-kernel/dkms ${BDEPEND} )"
+
+# @ECLASS_VARIABLE: DKMS_PACKAGES
+# @OUTPUT_VARIABLE
+# @DESCRIPTION:
+# After dkms_src_install or dkms_dopackage this array will be
+# populated with all dkms packages installed by the ebuild. The names
+# and versions of each package are separated with a ':'.
+DKMS_PACKAGES=()
+
+# @FUNCTION: dkms_gentoofy_conf
+# @USAGE: <list of files>
+# @DESCRIPTION:
+# Adds linux-mod-r1's MODULES_MAKEARGS and the ebuilds modargs to any
+# make calls in an existing dkms.conf. This function must be called
+# for every dkms.conf that will be installed to ensure that the users
+# compiler choice and flags are respected by DKMS at runtime.
+# Multiple files may be passed to this function as arguments. If no
+# arguments are given than this function runs on the dkms.conf in the
+# present working directory. Does nothing if USE=dkms is disabled.
+dkms_gentoofy_conf() {
+	debug-print-function ${FUNCNAME} "$@"
+
+	use dkms || return 0
+	[[ -z ${MODULES_OPTIONAL_IUSE} ]] ||
+		use "${MODULES_OPTIONAL_IUSE#+}" || return 0
+
+	local file input=( "${@}" )
+	[[ ${#} -eq 0 ]] && input=( dkms.conf )
+
+	# This will set edkmsargs
+	dkms_sanitize_makeargs
+
+	for file in "${input[@]}"; do
+		[[ -f ${file} ]] ||
+			die "${FUNCNAME}: DKMS conf does not exist: ${file}"
+		# Check if make command already respecting our options.
+		# If not, sed our options in for all MAKE[#].
```

Comment:
This is a bit gnarly. Is this a bug in dkms, or what? Or just people writing crappy Makefiles and we want to fix them up to work with dkms? I don't know we should have this workaround.

Reply:
It's just a check to prevent sed'ing in the `MODULES_MAKEARGS` more then once since this function is called for each module in the `modlist`.

Reply:
Oh, I see. But if it's not respecting it, why is this our problem? Why are we mangling it in the eclass at all?

Reply:
> Perhaps I could rewrite this a bit to take this part of the `if` outside of the `for` loop.

Actually no, we cannot do that since the `modlist` may contain modules in different source directories (`mod[1]`) and then we may create multiple `dkms.conf`s, each would have to be `dkms_gentoofy_conf`'ed if it exists already but if there are duplicate source dirs in the `modlist` then we do still need this check to verify that we have not already run `dkms_gentoofy_conf` on that `dkms.conf`.

Reply:
> This is a bit gnarly.

Restructured this a bit. Now we keep track of the `dkms.conf` files we touch, and then afterwards we call `dkms_gentoofy_conf` them outside of the `for` loop. This means we can now safely assume this function is only called once and we can therefore drop this weird condition. As an additional bonus, this also means we only have to call `dkms_sanitize_makeargs` in one place, which makes the structure as a whole look a bit cleaner. Let me know what you think!

Also rebased the PR.